Add support for Model Context Protocol

martin-fleck-at opened the pull request:
#### What it does

Registers GLSP-announced MCP servers with the IDE's MCP frontend
service so the agent can use them immediately after IDE startup.

MCP integration
- Discovers MCP server descriptors on each GLSP client contribution
  and adds them to the IDE's MCP host on startup.
- Triggers an explicit start since the host's autostart hook only
  fires at frontend init.
- Per-registration cleanup releases server entries and state-change
  subscriptions on GLSP failure or extension dispose, avoiding
  stale entries across IDE restarts under random-port defaults.

Adopters
- Set the configuration on the GLSP client contribution to enable
  MCP for a diagram language.
- The workflow example demonstrates the end-to-end wiring with a
  transient connect notification and a copy-URL action.

Misc
- Sends a shutdown notification before the GLSP connection drops
  so the server tears down cleanly with the parent session.

Part of https://github.com/eclipse-glsp/glsp/issues/1546
Requires https://github.com/eclipse-glsp/glsp-client/pull/456 and https://github.com/eclipse-glsp/glsp-server-node/pull/120

#### How to test

- Nothing should break in the default case
- To use the new mcp server, you need to use a GLSP server that supports it, i.e., the [GLSP Node Server](https://github.com/eclipse-glsp/glsp-server-node/pull/120)
- Startup the Workflow example
- Check the AI Configuration > MCP Server view that should show the GLSP MCP server correctly and allow you to connect to it and show all tools. Please note that there is no support for MCP resources in Theia yet.  

#### Follow-ups

<!-- Please list potential follow-up work, including known issues, possible future work, identified technical debt, and potentially introduced technical debt. If the PR introduces technical debt, specify the reason why this is acceptable. Please create tickets and link them here. Please use the label "technical debt" for new issues when it applies. -->

#### Changelog

<!-- Please check, when if it applies to your change. -->

- [x] This PR should be mentioned in the changelog
- [ ] This PR introduces a breaking change (if yes, provide more details below for the changelog and the migration guide)

martin-fleck-at committed:
feat(theia-integration): add Model Context Protocol integration

Registers GLSP-announced MCP servers with the IDE's MCP frontend
service so the agent can use them immediately after IDE startup.

MCP integration
- Discovers MCP server descriptors on each GLSP client contribution
  and adds them to the IDE's MCP host on startup.
- Triggers an explicit start since the host's autostart hook only
  fires at frontend init.
- Per-registration cleanup releases server entries and state-change
  subscriptions on GLSP failure or extension dispose, avoiding
  stale entries across IDE restarts under random-port defaults.

Adopters
- Set the configuration on the GLSP client contribution to enable
  MCP for a diagram language.
- The workflow example demonstrates the end-to-end wiring with a
  transient connect notification and a copy-URL action.

Misc
- Sends a shutdown notification before the GLSP connection drops
  so the server tears down cleanly with the parent session.

Part of https://github.com/eclipse-glsp/glsp/issues/1546

tortmayr commented on .github/workflows/e2e.yml:
Why was this added? 
Do we currently even have e2e tests for the mcp feature?

tortmayr commented on configs/local-linking.sh:
Not sure if it's worth maintaining this script any longer.
We now have linking functionality built in the glsp cli tool.

We should probably update our repos accordingly.
But that's a thing for a follow-up

tortmayr commented on examples/workflow-theia/src/browser/workflow-glsp-client-contribution.ts:
We could probably align this into one `getPortFromEnv()` method that just takes the env var as parameter
Instead of three more or less identical methods

tortmayr commented on packages/theia-integration/src/browser/glsp-client-contribution.ts:
If this.glspClient hasn't resolved yet when dispose fires (e.g., contribution disposed during startup), the await will block the async body indefinitely, and connection.dispose() / channel.dispose() will never execute.
  The previous code pushed connection and channel as separate disposables, guaranteeing synchronous cleanu

tortmayr commented on packages/theia-mcp-integration/package.json:
```suggestion
    "generate:index": "glsp generateIndex src/browser -s -f",
```

tortmayr commented on .claude/skills/generate-changelog/SKILL.md:
The CHANGELOG entry formatting changed from -    (dash + 3 spaces) to -  (dash + 1 space), which contradicts the style guide in SKILL.md line 84 that specifies "Entry prefix: -    (dash + exactly 3 spaces)". If this is an intentional style migration,
  the SKILL.md should be updated to match. If not, the CHANGELOG entry should retain the original format.

  This might be related to the underlying version of prettier/eslint/markdown-lint that is used. I also experienced similar issues in other glsp projects.

tortmayr commented on packages/theia-mcp-integration/src/browser/glsp-mcp-registration.ts:
Remove the override fluff. (everything after the ;)

tortmayr commented on packages/theia-mcp-integration/src/browser/glsp-mcp-registration.ts:
No need to add an additional comment for the general toDispose pattern

tortmayr commented on packages/theia-mcp-integration/src/browser/glsp-mcp-registration.ts:
Same. no need to comment dispose

tortmayr commented on packages/theia-mcp-integration/src/browser/glsp-mcp-registration.ts:
maybe remove comment here...

martin-fleck-at commented on packages/theia-integration/src/browser/glsp-client-contribution.ts:
Good catch! We do a best effort shutdown now if the client already resolved but otherwise we will simply dispose as before.

martin-fleck-at commented on examples/workflow-theia/src/browser/workflow-glsp-client-contribution.ts:
good idea!

martin-fleck-at commented on configs/local-linking.sh:
yeah, I think so too, should be done in a proper cleanup.

martin-fleck-at commented on .github/workflows/e2e.yml:
We do not, so I removed it again.

martin-fleck-at commented on .claude/skills/generate-changelog/SKILL.md:
The prettier actually enforces the 1 space when we do `yarn format`. I updated the skill for this as it proposes 3 spaces

martin-fleck-at committed:
PR review

- glsp-client-contribution: guard shutdown await + try/finally cleanup
- workflow-glsp-client-contribution: fold port-getters into helper
- glsp-mcp-registration: drop verbose JSDoc/comments
- theia-mcp-integration/package.json: drop dead generate:index args
- e2e.yml: drop GLSP_MCP_SERVER_PORT (no MCP e2e tests yet)
- generate-changelog SKILL: match prettier-enforced list style
- yarn.lock: bump GLSP packages to next.9, pull in server-mcp
- export: migrate off deprecated ExportSvgAction /
RequestExportSvgAction
- export: new TheiaExportCommandHandler picks SVG or PNG via QuickPick,
  uses request/response dispatch, writes PNG as base64-decoded
BinaryBuffer

tortmayr reviewed (approved):
LGTM! 👍🏼
